File: python/basics/string/392_is_subsequence.py
'''
https://leetcode.com/problems/is-subsequence/description/?envType=study-plan-v2&envId=top-interview-150

simple pointer stuff
'''

def is_subsequence(s: str, t: str):
    s_len = len(s)
    t_len = len(t)
    if s == None or len(s) == 0:
        return True
    i = 0
    j = 0
    while i < s_len and j < t_len:
        if s[i] == t[j]:
            i += 1
            if i == s_len:
                return True
        j += 1
    return False


s = "abc"
t = "ahbgdc"
#Output: true
ret = is_subsequence(s, t)
print("is_subsequence: " + str(ret))

s = "axc"
t = "ahbgdc"
#Output: false
ret = is_subsequence(s, t)
print("is_subsequence: " + str(ret))
 

File: problem_solving/2-pointer/cpp/392_is_subsequence.cpp
/*
https://leetcode.com/problems/is-subsequence/
 */

#include<string>
#include<iostream>
#include "../../utils/utils.h"


 bool is_subsequence(const std::string& s, const std::string& t) {
    if(s.empty()) {
        return true;
    }
    int k = 0;
    int len_t = t.length();
    int len_s = s.length();
    for (int i = 0; i < len_t; ++i) {
        if (t[i] == s[k]) {
            k += 1;
            if(k==len_s) {
                return true;
            }
        }
    }
    return false;
}


 int main() {
    std::string s = "abc";
    std::string t = "ahbgdc";
    //Output: true
    auto result = is_subsequence(s, t);
    std::cout << "result: " << result << std::endl;

    s = "";
    t = "ahbgdc";
    //Output: true
    result = is_subsequence(s, t);
    std::cout << "result: " << result << std::endl;

    s = "a";
    t = "c";
    //Output: false
    result = is_subsequence(s, t);
    std::cout << "result: " << result << std::endl;

    s = "axc";
    t = "ahbgdc";
    //Output: false
    result = is_subsequence(s, t);
    std::cout << "result: " << result << std::endl;
  
 }